File: app/src/main/java/com/gee12/mytetroid/views/adapters/TagsListAdapter.java
package com.gee12.mytetroid.views.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;

import com.gee12.mytetroid.R;
import com.gee12.mytetroid.SortHelper;
import com.gee12.mytetroid.model.TetroidRecord;
import com.gee12.mytetroid.model.TetroidTag;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public class TagsListAdapter extends BaseAdapter {

    /**
     *
     */
    private class TagsViewHolder {
        TextView nameView;
        TextView recordsCountView;
    }

    private final LayoutInflater mInflater;
    private final Context mContext;
    private ArrayList mData;

    public TagsListAdapter(Context context) {
        this.mContext = context;
        this.mInflater = (LayoutInflater)context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        this.mData = new ArrayList<>();
    }

    public void setDataItems(Map<String,TetroidTag> data, SortHelper sortHelper) {
        this.mData = new ArrayList<>(data.entrySet());
        sort(sortHelper);
        notifyDataSetChanged();
    }

    @Override
    public int getCount() {
        return mData.size();
    }

    @Override
    public Map.Entry<String, TetroidTag> getItem(int position) {
        return (Map.Entry) mData.get(position);
    }

    @Override
    public long getItemId(int position) {
        return mData.get(position).hashCode();
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        TagsViewHolder viewHolder;
        if (convertView == null) {
            viewHolder = new TagsViewHolder();
            convertView = mInflater.inflate(R.layout.list_item_tag, null);
            viewHolder.nameView = convertView.findViewById(R.id.tag_view_name);
            viewHolder.recordsCountView = convertView.findViewById(R.id.tag_view_records_count);
            convertView.setTag(viewHolder);
        } else {
            viewHolder = (TagsViewHolder) convertView.getTag();
        }

        Map.Entry<String, TetroidTag> item = getItem(position);
        // название метки
        viewHolder.nameView.setText(item.getKey());
        // количество записей
        final List<TetroidRecord> records = item.getValue().getRecords();
        viewHolder.recordsCountView.setText(String.format(Locale.getDefault(), "[%d]", records.size()));

        return convertView;
    }

    private void sort(SortHelper sortHelper) {
        if (sortHelper == null)
            return;
        sort(sortHelper.isByName(), sortHelper.isAscent());
    }

    public void sort(boolean byName, boolean isAscent) {
        Collections.sort(mData, new Comparator<Map.Entry<String, TetroidTag>>() {
            @Override
            public int compare(Map.Entry<String, TetroidTag> o1, Map.Entry<String, TetroidTag> o2) {
                if (o1 == o2) {
                    return 0;
                } else if (o1 == null) {
                    return (isAscent) ? 1 : -1;
                } else if (o2 == null) {
                    return (isAscent) ? -1 : 1;
                } else {
                    int res;
                    if (byName) {
                        res = o1.getKey().compareTo(o2.getKey());
                    } else {
                        res = Integer.compare(o1.getValue().getRecords().size(), o2.getValue().getRecords().size());
                        if (res == 0) {
                            res = o1.getKey().compareTo(o2.getKey());
                        }
                    }
                    return (isAscent) ? res : res * -1;
                }
            }
        });
        notifyDataSetChanged();
    }

}

File: app/src/main/java/com/gee12/mytetroid/ui/tag/TagsListAdapter.kt
package com.gee12.mytetroid.ui.tag

import android.content.Context
import android.graphics.Color
import android.text.Spannable
import android.text.SpannableString
import android.text.style.AbsoluteSizeSpan
import android.text.style.ForegroundColorSpan
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.gee12.mytetroid.R
import com.gee12.mytetroid.domain.SortHelper
import com.gee12.mytetroid.domain.provider.IResourcesProvider
import com.gee12.mytetroid.model.TetroidTag
import java.util.*

class TagsListAdapter(
    private val context: Context,
    private val resourcesProvider: IResourcesProvider,
    private val checkIsSelected: (TetroidTag) -> Boolean,
    private val onClick: (View, TetroidTag) -> Unit,
    private val onLongClick: (View, TetroidTag) -> Unit,
) : BaseAdapter() {

    private inner class TagsViewHolder {
        lateinit var nameCountView: TextView
    }

    private val inflater: LayoutInflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
    private var dataSet: MutableList<Map.Entry<String, TetroidTag>> = mutableListOf()

    fun reset() {
        dataSet = mutableListOf()
        notifyDataSetChanged()
    }

    fun setDataItems(data: Map<String, TetroidTag>, sortHelper: SortHelper?) {
        dataSet = ArrayList(data.entries)
        sort(sortHelper)
        notifyDataSetChanged()
    }

    override fun getCount(): Int {
        return dataSet.size
    }

    override fun getItem(position: Int): Map.Entry<String, TetroidTag> {
        return dataSet[position]
    }

    override fun getItemId(position: Int): Long {
        return dataSet[position].hashCode().toLong()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val view: View
        val viewHolder: TagsViewHolder
        if (convertView == null) {
            viewHolder = TagsViewHolder()
            view = inflater.inflate(R.layout.list_item_tag, null)
            viewHolder.nameCountView = view.findViewById(R.id.tag_view_name_count)
            view.tag = viewHolder
        } else {
            view = convertView
            viewHolder = convertView.tag as TagsViewHolder
        }
        val (key, tag) = getItem(position)

        val backgroundColor = if (checkIsSelected(tag)) {
            ContextCompat.getColor(context, R.color.colorSelectedTag)
        } else {
            Color.TRANSPARENT
        }
        view.setBackgroundColor(backgroundColor)

        val recordsCount = "[%d]".format(tag.records.size)

        if (tag.isEmpty) {
            val emptyTag = resourcesProvider.getString(R.string.title_empty_tag)
            val nameCount = SpannableString("$emptyTag $recordsCount")
            nameCount.setSpan(
                AbsoluteSizeSpan(context.resources.getDimension(R.dimen.font_small).toInt()),
                emptyTag.length + 1, nameCount.length,
                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            viewHolder.nameCountView.text = nameCount
            viewHolder.nameCountView.setTextColor(ContextCompat.getColor(context, R.color.colorLightText))
        } else {
            val nameCount = SpannableString("$key $recordsCount")
            nameCount.setSpan(
                ForegroundColorSpan(ContextCompat.getColor(context, R.color.colorBase2Text)),
                key.length + 1, nameCount.length,
                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            nameCount.setSpan(
                AbsoluteSizeSpan(context.resources.getDimension(R.dimen.font_small).toInt()),
                key.length + 1, nameCount.length,
                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            viewHolder.nameCountView.text = nameCount
            viewHolder.nameCountView.setTextColor(ContextCompat.getColor(context, R.color.colorBaseText))
        }

        view.setOnClickListener {
            onClick(view, tag)
        }

        view.setOnLongClickListener {
            onLongClick(view, tag)
            true
        }
        return view
    }

    private fun sort(sortHelper: SortHelper?) {
        if (sortHelper == null) return
        sort(sortHelper.isByName, sortHelper.isAscent)
    }

    fun sort(byName: Boolean, isAscent: Boolean) {
        Collections.sort(dataSet, Comparator { obj1, obj2 ->
            if (obj1 === obj2) {
                0
            } else if (obj1 == null) {
                if (isAscent) 1 else -1
            } else if (obj2 == null) {
                if (isAscent) -1 else 1
            } else {
                var res: Int
                if (byName) {
                    res = obj1.key.compareTo(obj2.key)
                } else {
                    res = obj1.value.records.size.compareTo(obj2.value.records.size)
                    if (res == 0) {
                        res = obj1.key.compareTo(obj2.key)
                    }
                }
                if (isAscent) res else res * -1
            }
        })
        notifyDataSetChanged()
    }
}